backend/test_observer/data_access/models.py

```diff
@@ -92,6 +97,30 @@ class Artefact(Base):
         UniqueConstraint("name", "version", "source", name="unique_artefact"),
     )
 
+    @validates("source")
+    def validate_source(
+        self, key: str, value: dict, family_name: str | None = None  # noqa: ARG002
+    ) -> dict:
+        """
+        Validate source field
+
+        :key: the key to validate. Not used now but is passed during the call.
+              I'm not replacing it with _ to avoid confusions in future
+        :value: the value to validate
+        :family_name: optional argument for the case when stage is not set yet
+                      but we need it to get family name
+        """
+        if (
+            self.stage and self.stage.family.name == FamilyName.SNAP.value
+        ) or family_name == FamilyName.SNAP.value:
+            # Check that store key is specified
+            if "store" not in value:
+                raise ValueError("Snap artefacts should have store key in source")
+            # Check that store key has a correct value
+            if not isinstance(value["store"], str):
+                raise ValueError("Store key in source field should be a string")
+        return value
```

Comment:
`validate_function` is needed when we create an artefact, since in the `validate_source`, the stage and stage_id are not defined yet, and we cannot check if the artefact belongs to snap family. But we can handle it in the `validate_artefact` function and create event listener to handle it
when an existing artefact is updated, the `validate_source` method is called and we already can get family there. that's why I keep both functions

Reply:
Oh I see your logic, that's why `family_name` can have type `None`. But I guess if we just make `validate_function` trigger on `before_update` too then there is no need for this right? It makes things simpler

Reply:
I've thought about that, but won't it slow the server?